web/ui_utils.js

```diff
@@ -318,44 +327,226 @@ function getVisibleElements(scrollEl, views, sortByVisibility = false) {
+  let percentVisible;
   let firstVisibleElementInd = views.length === 0 ? 0 :
-    binarySearchFirstItem(views, isElementBottomBelowViewTop);
+    binarySearchFirstItem(views, isElementBottomAfterViewTop);
+
+  if (views.length > 0) {
+    // In some combinations of page sizes and scrolling modes,
+    // isElementBottomAfterViewTop doesn't satisfy the binary search condition:
+    // there can be pages with bottoms above the view top between pages with
+    // bottoms below. This block detects and corrects that error, while trying
+    // to minimize the performance impact to the simple vertical scrolling case
+    // with pages stacked singly atop each other.
+
+    // The idea is that pages appear in rows, and we may need to backtrack to
+    // the beginning of either the found element's row or the previous row,
+    // depending on whether the current row intersects the top of the scroll
+    // pane. (We never need to backtrack further, since binary search would
+    // not have positioned us inside a fully visible row unless the previous
+    // row is not fully visible.)
+
+    let backtrack = false, pageTop, elt;
+    elt = views[firstVisibleElementInd].div;
+    pageTop = elt.offsetTop + elt.clientTop;
+    if (pageTop < top) {
+      // This is the simple case: the found page intersects the top of the
+      // scroll pane, so we backtrack to the first page in this page's row.
+      backtrack = true;
+    } else if (firstVisibleElementInd >= 2) {
+      // The found page does not intersect the top of the scroll pane, so thanks
+      // to the binary search, we know we're at the beginning of a row. Use the
+      // top of the previous page to detect the previous row.
+      elt = views[firstVisibleElementInd - 1].div;
+      pageTop = elt.offsetTop + elt.clientTop;
+
+      // If the page back two is above the previous page (as is the case in
+      // simple vertical scrolling), we don't need to backtrack. If the previous
+      // page had been visible, binary search wouldn't have placed us here.
+      elt = views[firstVisibleElementInd - 2].div;
+      if (elt.offsetTop + elt.clientTop + elt.clientHeight > pageTop) {
+        // But if that page is still in the previous row, we do backtrack to the
+        // beginning of that row.
+        firstVisibleElementInd -= 2;
+        backtrack = true;
+      }
+    }
+    if (backtrack) {
+      // Now we backtrack to the first page that still has its bottom below
+      // pageTop, which is the top of a page in the first visible row.
+      for (; firstVisibleElementInd > 0; --firstVisibleElementInd) {
+        elt = views[firstVisibleElementInd - 1].div;
+        if (elt.offsetTop + elt.clientTop + elt.clientHeight <= pageTop) {
+          // We hit this break immediately if pages are stacked vertically, so
+          // vertical scrolling stays fast (perhaps at the expense of other
+          // scrolling modes).
+          break;
+        }
+      }
+    }
+  }
```

Comment:
The primary reason (for this and other decisions made in this function) is to aggressively preempt any suspicions that this change would have a performance impact on the default scrolling mode. From @Snuffleupagus's previous review:
> I'm just not sure how we'll be able to address that, with only limited changes to `getVisibleElements`, such that we won't end up regressing performance for the regular (vertical) scrolling mode.

Since there aren't tests that I could benchmark to demonstrate that this change would not regress performance, I decided to err whenever I could on the side of making it obvious from looking at the code diff that the performance impact would be minimal—that meant no new function calls that weren't absolutely necessary, no run-time checking for the horizontal vs. vertical case, etc. I would be happy to reimplement this function with more of an eye toward clean code, but it would be trading off against the obviousness of no performance regressions. Let me know if that's something you want me to go forward with, but I would appreciate it if you can first do what you can to check that a reasonable (obviously this is subjective) rewrite wouldn't get blocked by someone being anxious about performance—I don't want to do a rewrite for nothing!

Reply:
The performance of code that touches DOM is generally DOM access. The overhead of JavaScript function calls is minimal.

As for performance: Previously, the function strictly used a binary search, so the performance would be rougly logarithmic. Now you're also doing backtracking, so the performance is potentially linear if a calculation went wrong. I haven't studied your algorithm at the algorthmic analysis level yet, I will still need to revisit that later (the inline comments look very clear, thanks for that).

If you want to measure: Since you've written this algorithm, you probably know best in which edge case the function would perform the worst. You could then create a PDF (or HTML document) that exhibits these changes, and use the browser devtools' Performance timeline (Firefox and Chrome both have these features) to check for any obvious performance issues.